src/runtime-rs/crates/resource/src/manager_inner.rs

```diff
@@ -71,12 +71,21 @@ impl ResourceManagerInner {
         for dc in device_configs {
             match dc {
                 ResourceConfig::ShareFs(c) => {
-                    let share_fs = share_fs::new(&self.sid, &c).context("new share fs")?;
-                    share_fs
-                        .setup_device_before_start_vm(self.hypervisor.as_ref())
-                        .await
-                        .context("setup share fs device before start vm")?;
-                    self.share_fs = Some(share_fs);
+                    self.share_fs = if self
+                        .hypervisor
+                        .capabilities()
+                        .await?
+                        .is_fs_sharing_supported()
+                    {
+                        let share_fs = share_fs::new(&self.sid, &c).context("new share fs")?;
+                        share_fs
+                            .setup_device_before_start_vm(self.hypervisor.as_ref())
+                            .await
+                            .context("setup share fs device before start vm")?;
+                        Some(share_fs)
+                    } else {
+                        None
```

Comment:
If SharerFS is not supported for the hypervisor, why not return an error?

Besides, I think this part is not related to the establishment of capabilities but the actual usage of capabilities in runtime, which should be  separated into other PR.

Reply:
As we are going to support multiple hypervisors, we need to consider about the use case of firecracker. It doesn't support using virtio-fs, we can only use it devicemapper snapshotter to achieve that.

Reply:
I agree with @Tim-0731-Hzt that the way the code is now we might not be able to return an error here.  However I also share @studychao 's intuition that returning an error would be the right thing to do.

I think the problem might be that we never check AFAICT whether `SharedFsInfo::shared_fs` is `Some()`.  If it is then I believe we should return an error here since that means that someone configured kata to request a shared fs but they're also using a hypervisor that doesn't support it.

In `VirtSandbox::prepare_for_start_sandbox()` we push a `ResourceConfig::ShareFs` instance onto `resource_configs` unconditionally.  I think that if we only did this when `hypervisor_config.shared_fs.shared_fs` was not `None` we could afford return an error here.  We could alternatively also do the check in `ResourceManagerInner` but I believe that that's unnecessarily late and if we want to do the check, then `VirtSandbox` is a clean place to do so.

Does that make sense, or am I completely off the mark?

Reply:
We do check `sharefs` is `Some`, [see this](https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/resource/src/volume/share_fs_volume.rs#L39). If the hypervisor doesn't support share-fs, we need to use `copy file` to copy the file into the guest.

Reply:
Right, so I'll have to apologise for a bad wording first - when I said "we never check whether `SharedFsInfo::shared_fs` is `Some()`" I really meant "we never check _before the code under review executes_". So you're right, `ShareFsVolume` does check, but I believe that it only runs later on, after `ResourceManagerInner::prepare_before_start_vm()` has exited, is that correct?  If so then its checking can't be helpful for `prepare_before_start_vm()`.

And please don't get me wrong, I'm not arguing that the code is wrong. I'm just trying to a) improve my understanding and b) make sure that the case where a user is running with a hypervisor that doesn't support shared fs but they also set `shared_fd = "virtiofs-<something>"' in their `configuration.toml` is handled reasonably. I believe it should fail early and with a good error message to which end I think it could be good if `SharedFsInfo::shared_fs` was also checked either here in `ResourceManagerInner::prepare_before_start_vm()` along with hypervisor capabilities, or in `VirtSandbox::prepare_for_start_sandbox()` along the line of

Reply:
Yes, it was checked when try to create a container. 
Yes, we need to check if user config `shared_fd = "virtiofs-<something>` for a hypervisor doesn't support shared-fs, but I think it is better to be handled by the configuration plugin, like [dragonball](https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-types/src/config/hypervisor/dragonball.rs) and [qemu](https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-types/src/config/hypervisor/qemu.rs) does.

Reply:
Alright, fair enough, thanks @Tim-0731-Hzt !